### ISSUE TITLE
[onert] Introduce `IGradientApplier`

### ISSUE THREAD

**Issue by ragmani** [Samsung/ONE#11394]
## What
Let's introduce `IGradientApplier` to fix `GradientAppiler` called early

## Why
I heard from @jyoungyun that `GradientApplier` is called early than computing gradients. This problem must be fixed.
And `GradientApplier` is now utilizing `ITrainableFunction`, but `ITrainableFunction` may cause `GradientApplier` to make more function calls than necessary. Also it seems better to separate the feature as an applier for memory optimization later.

## Tasks
- Introduce `IGradientApplier`
  - #11399
- Apply `IGradientApplier` to `train::GradientApplier`
  - #11419
  - #11477

Draft: #11396

**Comment by ragmani** [comment]
- Before draft
```bash
$ ./Product/x86_64-linux.release/out/bin/onert_train mnist_keras.circle --load_expected:raw out/train.output.1000.bin --load_input:raw out/train.input.1000.bin --data_length 1000 --learning_rate 0.01
Model Expected Filename out/train.output.1000.bin
Model Input Filename out/train.input.1000.bin
Model Filename mnist_keras.circle
Epoch 1/5 - 3.653ms/step - loss: [0] 14.9167
Epoch 2/5 - 3.633ms/step - loss: [0] 1.2761
Epoch 3/5 - 3.639ms/step - loss: [0] 0.8336
Epoch 4/5 - 3.642ms/step - loss: [0] 0.6277
Epoch 5/5 - 3.647ms/step - loss: [0] 0.5127
===================================
MODEL_LOAD   takes 0.293 ms
PREPARE      takes 1.785 ms
EXECUTE      takes 566.947 ms
- MEAN     :  566.947 ms
- MAX      :  566.947 ms
- MIN      :  566.947 ms
- GEOMEAN  :  566.947 ms
===================================
```

- After draft
```bash
$ ./Product/x86_64-linux.release/out/bin/onert_train mnist_keras.circle --load_expected:raw out/train.output.1000.bin --load_input:raw out/train.input.1000.bin --data_length 1000 --learning_rate 0.01
Model Expected Filename out/train.output.1000.bin
Model Input Filename out/train.input.1000.bin
Model Filename mnist_keras.circle
Epoch 1/5 - 3.690ms/step - loss: [0] 8.7752
Epoch 2/5 - 3.681ms/step - loss: [0] 1.8136
Epoch 3/5 - 3.681ms/step - loss: [0] 1.1870
Epoch 4/5 - 3.665ms/step - loss: [0] 0.8873
Epoch 5/5 - 3.680ms/step - loss: [0] 0.7104
===================================
MODEL_LOAD   takes 1.059 ms
PREPARE      takes 2.147 ms
EXECUTE      takes 572.137 ms
- MEAN     :  572.137 ms
- MAX      :  572.137 ms
- MIN      :  572.137 ms
- GEOMEAN  :  572.137 ms
===================================

```

**Comment by jyoungyun** [comment]
When I test draft code using 1 data, 1 batch, 5 epoch, the result is exactly the same with tensorflow result. :)

**tensorflow**
```
$ python3 tensorflow_run.py -m ./mnist_keras -i ../data/fashion_mnist/train.input.1.bin -l ../data/fashion_mnist/train.output.1.bin --data_length 1 --batch_size 1
Epoch 1/5
1/1 [==============================] - 0s 111ms/step - loss: 130.7952 - mean_squared_error: 130.7952
Epoch 2/5
1/1 [==============================] - 0s 2ms/step - loss: 77.1541 - mean_squared_error: 77.1541
Epoch 3/5
1/1 [==============================] - 0s 1ms/step - loss: 0.6691 - mean_squared_error: 0.6691
Epoch 4/5
1/1 [==============================] - 0s 1ms/step - loss: 0.2203 - mean_squared_error: 0.2203
Epoch 5/5
1/1 [==============================] - 0s 1ms/step - loss: 0.1271 - mean_squared_error: 0.1271
```

**onert_train**
```
$ ./Product/out/bin/onert_train --modelfile ../Personal/models/mnist_keras/mnist_keras/mnist_keras.circle --load_input:raw ../Personal/models/data/fashion_mnist/train.input.1.bin --load_expected:raw ../Personal/models/data/fashion_mnist/train.output.1.bin --data_length 1 --learning_rate 0.01 --epoch 5 --batch_size 1
Model Expected Filename ../Personal/models/data/fashion_mnist/train.output.1.bin
Model Input Filename ../Personal/models/data/fashion_mnist/train.input.1.bin
Model Filename ../Personal/models/mnist_keras/mnist_keras/mnist_keras.circle
Epoch 1/5 - time: 22.136ms/step - loss: [0] 130.7952
Epoch 2/5 - time: 22.093ms/step - loss: [0] 77.1541
Epoch 3/5 - time: 21.540ms/step - loss: [0] 0.6691
Epoch 4/5 - time: 21.414ms/step - loss: [0] 0.2203
Epoch 5/5 - time: 21.153ms/step - loss: [0] 0.1271
```

However, as with your test results, if there is a lot of data, there is still an error.
I will investigate this more.

**Comment by ragmani** [comment]
> However, as with your test results, if there is a lot of data, there is still an error.
I will investigate this more.

[My test results](https://github.com/Samsung/ONE/issues/11394#issuecomment-1698934910) are before and after fixing the problem of calling `GradientApplier` early, so they are different. Please notice that the loss values have improved.

**Comment by ragmani** [comment]
Done.
